### PR TITLE
Add third party invite support to spam checker

### PR DESCRIPTION
**Pull request by erikjohnston**
Currently we just always allow third party invites, which isn't what we want.

### CODE REVIEW

File: synapse/events/spamcheck.py [hunk 1]
```diff
@@ -46,14 +46,19 @@ def check_event_for_spam(self, event):
...
 
         If this method returns false, the invite will be rejected.
 
         Args:
             inviter_userid (str)
-            invitee_userid (str)
+            invitee_userid (str|None): The user ID of the invitee. Is None
+                if this is a third party invite and the 3PID is not bound to a
+                user ID.
+            invitee_userid (dict|None): If a third party invite then is a dict
```
Comment (richvdh) [inline synapse/events/spamcheck.py]: third_party_invite

File: synapse/handlers/room_member.py [hunk 2]
```diff
@@ -744,6 +747,20 @@ def do_3pid_invite(
...
+            third_party_invite={
+                "medium": medium,
+                "address": address,
+            },
+            room_id=room_id,
+            new_room=new_room,
+        ):
+            logger.info("Blocking invite due to spam checker")
+            raise SynapseError(
+                403, "Invites have been disabled on this server",
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: surely this is misleading?

File: synapse/rulecheck/domain_rule_checker.py [hunk 3]
```diff
@@ -62,19 +65,28 @@ def __init__(self, config):
...
+                        room_id, new_room):
         """Implements synapse.events.SpamChecker.user_may_invite
         """
         if self.can_only_invite_during_room_creation and not new_room:
             return False
 
+        if not self.can_invite_by_third_party_id and third_party_invite:
+            return False
+
+        if third_party_invite and not invitee_userid:
```
Comment (richvdh) [inline synapse/rulecheck/domain_rule_checker.py]: comment pls. what's this doing?

File: synapse/handlers/room_member.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/handlers/room_member.py]: This is in line with what we say elsewhere for banned invites, but yes. I don't know whether we want to change it now and what to change it to if we did.

Comment (richvdh) [reply synapse/handlers/room_member.py]: fair

**Pull request merged**